experiments/standalone/neural_snow/data_tools.jl

```diff
@@ -0,0 +1,362 @@
+            push!(rolleddata, row)
+        end
+    end
+    return stack2DF(rolleddata, names(data))
+end
+
+"""
+    snowsplit(air_temp, hum, precip)
+
+Engineer total water content of precipitation into snow and rain portions,
+accoridng to the paper outlined in https://www.nature.com/articles/s41467-018-03629-7.
+
+# Arguments
+- `air_temp::Vector{<:Real}`: the air temperature data.
+- `hum::Vector{<:Real}`: the relative humidity data.
+- `precip::Vector{<:Real}`: the precipitation data.
+"""
+function snowsplit(air_temp::Vector{<:Real}, hum::Vector{<:Real}, precip::Vector{<:Real})
+    α = -10.04
+    β = 1.41
+    γ = 0.09*100 #need to figure out if HUM in decimal 0-1 or 0-100 for this constant!
+    #snow_frac = round.(1.0 ./ (1.0 .+ exp.(α .+ β .* air_temp .+ γ .* hum))) #leave-be or round?
+    snow_frac = (1.0 ./ (1.0 .+ exp.(α .+ β .* air_temp .+ γ .* hum))) #leave-be or round?
+    dprecipdt_snow = snow_frac .* precip
+    dprecipdt_rain = (1.0 .- snow_frac) .* precip
+    return (dprecipdt_snow, dprecipdt_rain)
+end
+
+"""
+    filter_phys!(data; ϵ = 0.2)
+
+Filter unphysical/undesirable data points out of the dataset.
+**Note: requires column names to match those of the paper for usage.
+
+# Arguments
+- `data::DataFrame`: the cleaned and processed data.
+- `ϵ::Real`: a filtering threshold, set to 0.5 cm.
+"""
+function filter_phys!(data::DataFrame;
+    eps::Real = 0.005)
+    #these need to be modified according to paper commentary and correct units.
+    zero_condition1 = (data[!, :SWE] .< eps) .& (data[!, :z] .< eps) .& (data[!, :dprecipdt] .< eps/86400.0)
+    data = data[Not(zero_condition1), :]
+    zero_condition2 = (data[!, :z] .!= 0.0f0) .& (data[!, :SWE] .== 0.0f0)
+    data = data[Not(zero_condition2), :]
+    zero_condition3 = (data[!, :dzdt] .> 0.0) .& (data[!, :dprecipdt] .== 0.0)
+    data = data[Not(zero_condition3), :]
+    data = data[data[!, :dSWEdt].<=data[!, :dprecipdt], :]
+    data = data[data[!, :SWE].<=data[!, :z], :]
+    #zero_condition_temp = (abs.(data[!, :dzdt]) .< 2 * eps) .& (data[!, :air_temp_avg] .> 9.0)
+    #data = data[Not(zero_condition_temp), :]
+end
+
+"""
+    prep_data(data; extract_vars, make_snow_split, physical_filter, ϵ)
+
+Prepare a cleaned data stream for model usage.
+**Note: Requires column names to match those of the paper for usage.
+
+# Arguments
+- `data::DataFrame`: the cleaned and processed data.
+- `extract_vars::Vector{Symbol}`: The list of columns to be used in the model. Default is the variables used in the paper.
+- `make_snow_split::Bool`: Boolean indicating whether to split precipitation into water and snow. Default is true.
+- `physical_filter::Bool`: Boolean indicating whether to filter data using filter_phys!. Default is true.
+- `ϵ::Real`: a filtering threshold, set to 0.5 cm.
+"""
+function prep_data(data::DataFrame;
+    extract_vars::Vector{Symbol}=[:date, :id, :z, :SWE, :rel_hum_avg, :sol_rad_avg, :wind_speed_avg, :air_temp_avg, :dprecipdt_snow, :dzdt],
```

Comment:
is this needed in this function? why not restrict to those columns earlier?

Reply:
This one is used usually independent of files that use the rest of datatools.jl, it is usually the first/only call made to an already-complete data csv file (see the test_run.jl file), so that is the earliest restriction.